src/crimson/os/seastore/seastore_types.cc

```diff
@@ -129,6 +129,81 @@ std::ostream &operator<<(std::ostream &out, const paddr_t &rhs)
   return out << ">";
 }
 
+journal_seq_t journal_seq_t::add_offset(
+      journal_type_t type,
+      seastore_off_t off,
+      seastore_off_t roll_start,
+      seastore_off_t roll_size) const
+{
+  assert(offset.is_absolute());
+  assert(off != MIN_SEG_OFF);
+  assert(roll_start >= 0);
+  assert(roll_size > 0);
+
+  segment_seq_t jseq = segment_seq;
+  seastore_off_t joff;
+  if (type == journal_type_t::SEGMENTED) {
+    joff = offset.as_seg_paddr().get_segment_off();
+  } else {
+    assert(type == journal_type_t::CIRCULAR);
+    auto boff = offset.as_blk_paddr().get_block_off();
+    assert(boff <= MAX_SEG_OFF);
+    joff = boff;
+  }
+  auto roll_end = roll_start + roll_size;
+  assert(joff >= roll_start);
+  assert(joff <= roll_end);
+
+  if (off >= 0) {
+    jseq += (off / roll_size);
+    joff += (off % roll_size);
+    if (joff >= roll_end) {
+      ++jseq;
+      joff -= roll_size;
+    }
+  } else {
+    auto mod = static_cast<segment_seq_t>((-off) / roll_size);
+    joff -= ((-off) % roll_size);
```

Comment:
If offset is blk_paddr_t, we probably need to use int64_t, not int32_t.

Reply:
If 2GB is not enough for journaling with RBM, I think we will need a separate PR to address this by introducing the original int32_t segment_off_t back, use it where appropriate, and then extend seastore_off_t to int64_t.


Refer to the commit https://github.com/ceph/ceph/commit/c93d0b7b1e7be2e3879ec384a951edb247a9f775